### PR TITLE
Fixes part of #4120, part of #1051: Fix a lot of build-time warnings

### PR DESCRIPTION
**Pull request by BenHenning**
## Explanation
Fixes part of #4120
Fixes part of #1051

Similar to #5400, this brings forward changes that would otherwise go in #4937 to simplify the transition to Kotlin 1.6.

Part of #4937 is introducing warnings-as-errors for both Kotlin and Java in order to reduce developer error, simplify the codebase, and minimize warnings when building (which can result in developer habits of ignoring warnings that might have real consequences to end users of the app). In order to keep the main migration PR smaller, this PR fixes all existing warnings and any new ones detected with the Kotlin 1.6 compiler that are not tied to Kotlin 1.5/1.6 API changes (those are part of #4937, instead). Fortunately, most of the changes could be brought forward into this PR.

Specific things to note:
- A few new issues were filed for SDK 33 deprecations caused, but not noted by, #5222): #5404, #5405, and #5406 and corresponding TODOs added. This PR opts for TODOs over actual fixes to minimize the amount of manual verification needed, and to try and keep the PR more focused on non-functional refactor changes (to reduce the risk as reverting this PR may be difficult if an issue is introduced).
- A lot of the fixes were removing redundant casts or null checks.
- The old mechanism we used for view models is deprecated, and had a lot of problems  (partially documented in #1051). This PR moves the codebase over to directly injecting view models instead of using the view model provider (thus getting rid of true Jetpack view models entirely in the codebase).
  - We never used the Jetpack functionality, and we were leaking a lot of context objects that could theoretically result in memory leaks.
  - The migration of view models in this way has already been ongoing in the codebase; this PR just finishes moving the rest of them over to remove the deprecated JetPack view model reference.
  - Note that this doesn't actually change the scope of the view models, and in fact they should largely behave as they always have.
  - ``ObservableViewModel`` was subsequently updated, and may be something we could remove in the future now that it's no longer a Jetpack view model.
  - The old view model binding code was removed, along with its test file exemptions. It's no longer used now that the view models have been finished being migrated over to direct injection.
- Some of the binding adapters didn't correctly correspond to their namespaced properties. I _think_ that the databinding compiler was still hooking them up correctly, but they produced build warnings that have now been addressed (specifically, 'app' is implied). Some other properties were using unusual namespaces, so these were replaced with 'app' versions for consistency & correctness.
- Some cases where SAM interfaces could be converted to lambdas were also addressed (mainly for ``Observer`` callbacks in UI code).
- ``DrawerLayout.setDrawerListener`` was replaced with calls to ``DrawerLayout.addDrawerListener`` since the former [is deprecated](https://developer.android.com/reference/androidx/drawerlayout/widget/DrawerLayout#setDrawerListener(androidx.drawerlayout.widget.DrawerLayout.DrawerListener)). This isn't expected to have a functional difference.
- Some other minor control flow warnings were addressed (such as dead code paths).
- ``when`` cases were updated to be comprehensive (as this is enforced starting in newer versions of Kotlin even for non-result based ``when`` statements).
- Some unused variables were removed and/or replaced with ``_`` per Kotlin convention.
- Some parameter names needed to be updated to match their override signatures.
- One change in ``ExitSurveyConfirmationDialogFragment`` involved removing parsing a profile ID. Technically this is a semantic change since now a crash isn't going to happen if the profile ID is missing or incorrect, but that seems fine since the fragment doesn't even need a profile ID to be passed.
- Some of the test activities were updated to bind a ``Runnable`` callback rather than binding to a method (just to avoid passing the unused ``View`` parameter and to keep things a bit simple binding-wise).
- Some cases were fixed where variables were being shadowed due to reused names in deeper scopes.
- There were some typing issues going on in tests with custom test application components. This has been fixed by explicitly declaring the application component types rather than them being implicit within the generated Dagger code.
- ``getDrawable`` calls were updated to pass in a ``Theme`` since the non-theme version is deprecated.
- Some Java property references were updated, too (i.e. using property syntax instead of Java getters when referencing Java code in Kotlin).
- In some cases, deprecated APIs were suppressed since they're needed for testing purposes.
- Mockito's ``verifyZeroInteractions`` has been deprecated in favor of ``verifyNoMoreInteractions``, so updates were made in tests accordingly.
- ``ExperimentalCoroutinesApi`` and ``InternalCoroutinesApi`` have been deprecated in favor of a newer ``OptIn`` method (which can actually be done via kotlinc arguments, but not in this PR). Thus, they've been outright removed in cases where not needed, and otherwise migrated to the ``OptIn`` approach where they do need to be declared.
- In some cases, Kotlin recommends using a ``toSet()`` conversion for iterable operations when it's more performant, so some of those cases (where noticed) have been addressed.
- Some unused parameter cases needed to be suppressed for situations when Robolectric is using reflection to access them.
- In some cases Android Studio would recommend transformation chain simplifications; these were adopted where obvious.
- There are a few new TODOs added on #3616 as well, to clean up deprecated references that have been suppressed in this PR.
- ``BundleExtensions`` was updated to implement its own version of the type-based ``getSerializable`` until such time as ``BundleCompat`` can be used, instead (per #5405).
- A **lot** of nullability improvements needed to happen throughout the JSON asset loading process since there was a lot of loose typing happening there.
- Some Kotlin & OkHttp deprecated API references were also updated to use their non-deprecated replacements.
- ``NetworkLoggingInterceptorTest`` was majorly reworked to ensure that the assertions would actually run (``runBlockingTest`` was being used which is deprecated, and something I try to avoid since it's very difficult to write tests that use it correctly). My investigations showed that the assertions weren't being called, so these tests would never fail. The new versions will always run the assertions or fail before reaching them, and fortunately the code under test passes the assertions correctly. Ditto for ``ConsoleLoggerTest``.
- Some parts of ``SurveyProgressController`` were reworked to have better typing management and to reduce the need for nullability management.
- Some generic typing trickiness needed to be fixed ahead of the Kotlin version upgrade in ``UrlImageParser``. See file comments & links in those comments for more context.
- ``BundleExtensionsTest`` had to be changed since ``getSerializableExtra`` is now deprecated. We also can't update the test to run SDK 33 since that requires upgrading Robolectric, and Robolectric can't be upgraded without upgrading other dependencies that eventually lead to needing to upgrade both Kotlin and Bazel (so it's a non-starter; this is a workaround until we can actually move to a newer version of Robolectric).
- There was some minor code-deduplication & cleanup done in ``ClickableAreasImage``.
- Some incorrect comments were removed in tests (to the effect of "using not-allowed-listed variables should result in a failure."). These seemed to have been copied from an earlier test, but the later tests weren't actually verifying that behavior so the comment wasn't correct.
- An unused method was removed from ``ConceptCardRetriever`` (``createWrittenTranslationFromJson``) and some other small cleanup/consolidation work happened in that class.
- Some stylistic changes were done in ``TopicController`` for JSON loading to better manage nullable situations, and to try and make the JSON loading code slightly more Kotlin idiomatic.

Note that overall the PR has relied **heavily** on tooling to detect warnings to fix, and automated tests to verify that the changes have no side effects.

Note also that this PR does not actually enable warnings-as-errors; that will happen in a downstream PR.

## Essential Checklist
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
N/A -- While this changes UI code, it should change very few UI behaviors and only failure cases for those it does affect. It's largely infrastructural-only and falls mainly under refactoring/cleanup work.

### CODE REVIEW

**Review by BenHenning — COMMENTED** [review]
Self-reviewed all changes.

**Comment by BenHenning** [comment]
PTAL @adhiamboperes.

Note that this PR is pretty straightforward to review, but it's very large and caused my computer to have some issues when trying to view the actual changes. It probably is worth breaking this into chunks when reviewing it.

I think it'd also be completely fine to delegate review of most or all of it to other people. I've self-reviewed the changes and find them non-concerning, but I defer to whatever you want to do as the primary reviewer & approver of the PR.

**Review by BenHenning — COMMENTED** [review]
Self-reviewed latest changes (pulled in some more fixes from #4886).

**Review by BenHenning — COMMENTED** [review]
Acking latest changes were merged cleanly.

**Review by BenHenning — COMMENTED** [review]
Latest merge commits went in without changes.

File: utility/src/main/java/org/oppia/android/util/parser/image/UrlImageParser.kt
Comment (adhiamboperes) [inline utility/src/main/java/org/oppia/android/util/parser/image/UrlImageParser.kt]: Why do we need to annotate this usage of `getOpacity()` when it is already annotated as deprecated in the `Drawable` class?

File: data/src/test/java/org/oppia/android/data/backends/gae/NetworkLoggingInterceptorTest.kt
Comment (adhiamboperes) [inline data/src/test/java/org/oppia/android/data/backends/gae/NetworkLoggingInterceptorTest.kt:62]: Why do we usually insist on formatting @inject fields on a single line when AS or ktlint formats them otherwise?

File: domain/src/main/java/org/oppia/android/domain/survey/SurveyProgressController.kt
Comment (adhiamboperes) [inline domain/src/main/java/org/oppia/android/domain/survey/SurveyProgressController.kt:525]: This is a nice improvement. I did have a hard time coming up with the suitable typing vs nullability management at the time of writing this code.

File: scripts/src/java/org/oppia/android/scripts/testing/TestBazelWorkspace.kt
Comment (adhiamboperes) [inline scripts/src/java/org/oppia/android/scripts/testing/TestBazelWorkspace.kt:249]: Is this a test-specific upgrade?

**Review by adhiamboperes — COMMENTED** [review]
Thanks @BenHenning!
I noted the following:

1. You removed the following comment from a number of files, but it isn't clear why: `// Using not-allowed-listed variables should result in a failure.`
2. There are undocumented changes in [ConceptCardRetriever.kt](https://github.com/oppia/oppia-android/pull/5402/files#diff-f037b8ecc0269cc34361ae2838b18171794645256957f685631c3476164ca728) and [TopicController.kt](https://github.com/oppia/oppia-android/pull/5402/files#diff-0943fb0fc5b908e5ed8da64f53610e2c83c88815de2c2a253c89505da56485ea). I think it is worth including the reasons in the description.

I also left a couple of comments in-line.

File: utility/src/main/java/org/oppia/android/util/parser/image/UrlImageParser.kt
Comment (BenHenning) [reply utility/src/main/java/org/oppia/android/util/parser/image/UrlImageParser.kt]: Hmm, I'm actually not sure. I removed this line and it still builds, so it's possible it was needed earlier before other changes were in place, or it's a warning that came from Android Studio and not the Kotlin compiler.

File: data/src/test/java/org/oppia/android/data/backends/gae/NetworkLoggingInterceptorTest.kt
Comment (BenHenning) [reply data/src/test/java/org/oppia/android/data/backends/gae/NetworkLoggingInterceptorTest.kt:62]: I prefer the format because it's less lines & easier to read. Eventually we'll be moving over to ktfmt which will hard enforce an exact style #2799 since ktlint is a bit too permissive in my opinion (I'd rather try and make all code look as though it was written by the same author).

There's not a lot we can do about Android Studio's formatting, unfortunately. I haven't found great ways of configuring it to actually match ktlint or ktfmt, but perhaps there are plugins that can help with his (I haven't looked in a long time).

FWIW from a reviewer side, I tend to only ask people to change it if the existing code is already single line (for consistency).

File: domain/src/main/java/org/oppia/android/domain/survey/SurveyProgressController.kt
Comment (BenHenning) [reply domain/src/main/java/org/oppia/android/domain/survey/SurveyProgressController.kt:525]: Thanks--I'm not sure it's better from a readability perspective though. Lambdas and function references seem to often decrease the comprehensibility of code, so I was a bit on the fence with generalizing things as was done here.

File: scripts/src/java/org/oppia/android/scripts/testing/TestBazelWorkspace.kt
Comment (BenHenning) [reply scripts/src/java/org/oppia/android/scripts/testing/TestBazelWorkspace.kt:249]: Yeah. A lot of this code was originally written in #4937 and haphazardly copied to this PR, and that PR's code originated in #4886. The Bazel PR updates rules_kotlin to 1.7.1 and I figured we can just keep that bit here even though it's not in exact alignment for a few releases. It doesn't make much of a difference in the build graph between tests that use this utility and production.

**Review by BenHenning — COMMENTED** [review]
Self-reviewed latest changes. No specific concerns.

I left replies to your comments and added one change to address one of them, plus pulled in the latest develop changes. PTAL @adhiamboperes.

File: scripts/src/java/org/oppia/android/scripts/testing/TestBazelWorkspace.kt
Comment (adhiamboperes) [reply scripts/src/java/org/oppia/android/scripts/testing/TestBazelWorkspace.kt:249]: Sounds good.

**Review by adhiamboperes — APPROVED** [review]
Everything looks good,  thanks @BenHenning!

**Comment by BenHenning** [comment]
@seanlip PTAL for owners-specific changes.

**Comment by seanlip** [comment]
@BenHenning Owners approval does not seem to be required for this PR, unless I'm missing something?

**Comment by BenHenning** [comment]
Hmm not sure why I saw that originally @seanlip. Thanks for double checking!

Thanks @adhiamboperes for the review!

Enabling auto-merge now that this is up-to-date with the latest develop changes.

**Review by BenHenning — COMMENTED** [review]
Reviewed latest merge. No conflicts so the merge went in cleanly.

**Pull request merged**
